protocols/dcutr/src/handler/relayed.rs

```diff
@@ -0,0 +1,380 @@
+                    Event::InboundNegotiationFailed {
+                        error: ProtocolsHandlerUpgrErr::Upgrade(UpgradeError::Select(
+                            NegotiationError::Failed,
+                        )),
+                    },
+                ));
+            }
+            _ => {
+                // Anything else is considered a fatal error or misbehaviour of
+                // the remote peer and results in closing the connection.
+                self.pending_error = Some(error.map_upgrade_err(|e| {
+                    e.map_err(|e| match e {
+                        EitherError::A(e) => EitherError::A(e),
+                        EitherError::B(v) => void::unreachable(v),
+                    })
+                }));
+            }
+        }
+    }
+
+    fn inject_dial_upgrade_error(
+        &mut self,
+        _open_info: Self::OutboundOpenInfo,
+        error: ProtocolsHandlerUpgrErr<<Self::OutboundProtocol as OutboundUpgradeSend>::Error>,
+    ) {
+        self.keep_alive = KeepAlive::No;
+
+        match error {
+            ProtocolsHandlerUpgrErr::Timeout => {
+                self.queued_events.push_back(ProtocolsHandlerEvent::Custom(
+                    Event::OutboundNegotiationFailed {
+                        error: ProtocolsHandlerUpgrErr::Timeout,
+                    },
+                ));
+            }
+            ProtocolsHandlerUpgrErr::Upgrade(UpgradeError::Select(NegotiationError::Failed)) => {
+                // The remote merely doesn't support the DCUtR protocol.
+                // This is no reason to close the connection, which may
+                // successfully communicate with other protocols already.
+                self.queued_events.push_back(ProtocolsHandlerEvent::Custom(
+                    Event::OutboundNegotiationFailed {
+                        error: ProtocolsHandlerUpgrErr::Upgrade(UpgradeError::Select(
+                            NegotiationError::Failed,
+                        )),
+                    },
+                ));
+            }
+            _ => {
+                // Anything else is considered a fatal error or misbehaviour of
+                // the remote peer and results in closing the connection.
+                self.pending_error =
+                    Some(error.map_upgrade_err(|e| e.map_err(|e| EitherError::B(e))));
+            }
+        }
+    }
+
+    fn connection_keep_alive(&self) -> KeepAlive {
+        self.keep_alive
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ProtocolsHandlerEvent<
+            Self::OutboundProtocol,
+            Self::OutboundOpenInfo,
+            Self::OutEvent,
+            Self::Error,
+        >,
+    > {
+        // Check for a pending (fatal) error.
+        if let Some(err) = self.pending_error.take() {
+            // The handler will not be polled again by the `Swarm`.
+            return Poll::Ready(ProtocolsHandlerEvent::Close(err));
```

Comment:
Yep, this will close the socket AFAIK.

Reply:
:+1: reference to the `enum`:

https://github.com/libp2p/rust-libp2p/blob/ea180aa3a88641f74f9fa7f89d6efdf48509892d/swarm/src/protocols_handler.rs#L311-L321

Would a comment have helped here? Grateful for suggestions.

Reply:
yep I saw that. I just wasn't sure if it meant (in practice) the socket connection or the substream. I know this is tricky since it may not even be a socket connection. Not sure what else we could add, but maybe something like:

```
/// Close the connection for the given reason. Note this will affect all handlers that use this connection.
```

Reply:
What do you think of https://github.com/libp2p/rust-libp2p/pull/2470 @MarcoPolo?

Reply:
With https://github.com/libp2p/rust-libp2p/pull/2470 merged, I am resolving here.